### ISSUE TITLE
Authentication Using External IDP

### ISSUE THREAD

**Issue by luke-gerschwitz** [mlflow/mlflow#10922]
Hi team,

I'm wondering how to achieve authentication in MLflow using an external IDP via OIDC/OAuth. Are there any existing plugins that supports this? And is this even feasible via a custom authentication plugin?

Additionally, once authentication is enabled, are we still able to manage existing permissions? As far as I know, the permissions are stored in basic_auth.db and there is no way to define/manage these outside of a standard database.

**Comment by BenWilson2** [comment]
We're looking into supporting OAuth2 in this quarter. Please stay tuned! Right now it's not possible to do what you're trying to do.

**Comment by kcfigaro** [comment]
Regarding OIDC/Oauth, you can easily achieve that by using oauth2-proxy as a sidecar pattern if you are serving a container-based tracking server. Here is my demo code https://github.com/cloudacode/mlflow-oauth-sidecar

**Comment by kharkevich** [comment]
I invented a new MLFlow plugin here: https://github.com/data-platform-hq/mlflow-oidc-auth, this plugin is based on the basic-auth MLFlow plugin, but supports OIDC/OAuth2 for GUI Auth and utilizes tokens (just basic auth) for non-GUI access (from notebook, pipeline, etc) at the same time.

@BenWilson2  please let me know if you like this idea and want to have something inside of MLFlow instead of 3rd party plugin

**Comment by anxietymonger** [comment]
Any update on support of OAuth? Thank you.

**Comment by vassav-nn** [comment]
> I invented a new MLFlow plugin here: https://github.com/data-platform-hq/mlflow-oidc-auth, this plugin is based on the basic-auth MLFlow plugin, but supports OIDC/OAuth2 for GUI Auth and utilizes tokens (just basic auth) for non-GUI access (from notebook, pipeline, etc) at the same time.
> 
> @BenWilson2 please let me know if you like this idea and want to have something inside of MLFlow instead of 3rd party plugin

Nice plugin, but no example how to use access token from create_access_token() in curl/python

Bug in callback(): if not any fails on big list in group_data
Fix 
        if not any(
            group["displayName"] == AppConfig.get_property("OIDC_GROUP_NAME")
            or group["displayName"] == AppConfig.get_property("OIDC_ADMIN_GROUP_NAME")
            for group in group_data["value"]
        ):
        
to 
        if (group["displayName"] == AppConfig.get_property("OIDC_GROUP_NAME") or group["displayName"] == AppConfig.get_property("OIDC_ADMIN_GROUP_NAME")) in group_data["value"]
        ):

**Comment by AlexMidili** [comment]
> We're looking into supporting OAuth2 in this quarter. Please stay tuned! Right now it's not possible to do what you're trying to do.


Greetings with all due respect, any update about OAuth2?

**Comment by luke-gerschwitz** [comment]
As a small update to anyone still seeking OAuth capabilities, it's worth investigating if AWS' managed MLflow fits your needs. It is accessible through and integrated with SageMaker. This way the authN/authZ can be offloaded to AWS. 

https://aws.amazon.com/about-aws/whats-new/2024/06/amazon-sagemaker-mlflow-capability/

**Comment by MehdiTantaoui-99** [comment]
Hi MLFlow team, it would be really great if we can have the OIDC authentication feature.

**Comment by hahahannes** [comment]
Hi everyone! There is already a nice third party plugin which implements several features at https://github.com/data-platform-hq/mlflow-oidc-auth

Features:
- API authentication based on OIDC token as well as long term API token
- Role based authorization for models and experiments  
- Control which user is allowed to use MlFlow based on the existence of a custom field in the OIDC token
- User registration via OIDC login 

Please react to this comment in case this might be interesting for the community!
(@Yc-Chen @kharkevich , @MehdiTantaoui-99 , @AlexMidili, @anxietymonger , @luke-gerschwitz, @BenWilson2 )

**Comment by BenWilson2** [comment]
@hahahannes this is really well done. We should definitely be featuring this in our docs. Furthermore, what is your appetite for having a plugin like this officially hosted within the mlflow domain?

**Comment by hahahannes** [comment]
Hi @BenWilson2 
Thank you very much for the fast response! Much appreciated.
I am not the maintainer of the project though but I really like it and I think the community would profit from it. I guess a first step could be to mention it in the docs for the auth plugins. And then maybe iterating on a possible merge? 

By "hosting within the mlflow domain" you mean hosting the repository under the MlFlow repository and keeping it as a seperate plugin or merging the code upstream?

**Comment by anxietymonger** [comment]
@Yc-Chen May I ask how did you handle auth over MLflow SDK (the python API) with oauth2-proxy? I knew it's a good tool for browser traffic, but had hard time dealing with the SDK.

**Comment by kcfigaro** [comment]
> @Yc-Chen May I ask how did you handle auth over MLflow SDK (the python API) with oauth2-proxy? I knew it's a good tool for browser traffic, but had hard time dealing with the SDK.

You only need to add your oauth client, which generates the TOKEN(JWT) for your machine, as an extra JWT issuer ID and set the TOKEN to the `MLFLOW_TRACKING_TOKEN` environment value. then MLflow SDK will request authentication through oauth2-proxy with the TOKEN.

You can find my completed example in my previous comment above.

**Comment by anxietymonger** [comment]
@kcfigaro Thanks for the reply. I've seen the example but my only concern is that do we have any proper way of getting the token rather than letting users (most of the cases data scientists who may not familiar with the web browser developer tools) retrieve the token through browsers? It's definitely a good and quick hack for a small sized team but not "proper" enough for  an enterprise-level application IMO.

**Comment by Yc-Chen** [comment]
@anxietymonger Yes, after comparing options, I ended up using 'htpasswd' option for SDK access. Then you will need to configure `MLFLOW_TRACKING_PASSWORD` and `MLFLOW_TRACKING_USERNAME` to use the basic authentication. Otherwise you will need to create jwt and handle token refresh. You may take a look at [AuthLib](https://docs.authlib.org/en/latest/).

**Comment by anxietymonger** [comment]
Thanks for sharing, folks. BTW, I currently end up using basic auth as well (with an nginx as a reverse proxy and auth server), while seeking for better solutions in the meantime.

**Comment by kcfigaro** [comment]
@anxietymonger May I ask how your other apps programmatically fetch the token? Essentially, OAuth2-Proxy simply relays the authentication request—more precisely, it retrieves the JSON Web Key from the IdP and verifies the JWT—rather than generating the token itself. If your IdP offers an SDK or CLI to fetch the token, you can use that to obtain the token and set it as an env variable to access MLflow over OAuth2-Proxy.

For example, if Google is your external IdP, you can use the Google SDK to fetch the token and programmatically pass it to your MLflow tracking code.

**Comment by anxietymonger** [comment]
Oh, didn't know google offers SDK for doing that. I'll try it out, many thanks!

For others who may be confused as me, the easiest way for getting google oauth token is via google cloud CLI: https://cloud.google.com/sdk/gcloud/reference/auth/print-access-token?hl=en

**Comment by kcfigaro** [comment]
Yes. that's correct and also there is out-of-box SDK available for programmatic access. so I would say this approach is "proper" enough for an enterprise-level application from the user experience perspective.

**Comment by dillonmabry** [comment]
Is there any future plan for this? Was curious since it's a very commonly requested feature. Much thanks on any info.

**Comment by BorjaArroyo** [comment]
Hi, is there any update in this topic?

**Comment by hahahannes** [comment]
Hi @BorjaArroyo @dillonmabry We are actively working on submitting a third party plugin to be incorporated to MlFlow for this :)

**Comment by ZisisFl** [comment]
Hello @hahahannes are there any updates on this?

**Comment by flaccid** [comment]
Requesting an update on this matter also please.

I can confirm that https://pypi.org/project/mlflow-oidc-auth/ (https://pypi.org/project/mlflow-oidc-auth/) works well using container image, https://github.com/flaccid/container-mlflow/blob/main/Dockerfile. My IdP is Okta.

In my case, I used https://github.com/bitnami/charts/tree/main/bitnami/mlflow in kubernetes with this setup and container arg `--app-name=oidc-auth` to use the plugin.

Seems that perhaps this is just a matter of asking the plugin authors if they want to contrib the plugin?

**Comment by ZisisFl** [comment]
Thank you @flaccid for your detailed response this is very nice as I use the Bitnami chart as well, I will give it a try for sure!
